File: mobile/src/engine/runner.ts
/**
 * The trading loop.
 *
 * The only place the components meet. It owns no strategy logic, no risk rules
 * and no venue detail — it moves data across the seams in a fixed order:
 *
 *   market data -> strategy -> position diff -> risk -> execution -> journal -> listeners
 *
 * Risk sits between the strategy and the venue so no signal reaches a venue
 * unchecked, and the journal is written before listeners are notified so a
 * process death cannot lose a fill that already happened.
 */

import type { EngineSettings } from './config';
import { Journal, today } from './journal';
import type { MarketDataSource } from './marketdata';
import { createMarketData } from './marketdata';
import type { Candle, Exposure, Fill, Order, Position, Signal } from './models';
import { newClientId } from './models';
import { RiskEngine, type RiskState } from './risk';
import { EmaCrossStrategy, type Strategy } from './strategy';
import { ExecutionError, PaperVenue } from './venue';

const KILL_KEY = 'kill_switch';
const PAUSE_KEY = 'paused';

export interface EngineStatus {
  running: boolean;
  killed: boolean;
  paused: boolean;
  mode: 'paper';
  venue: string;
  exchange: string;
  symbol: string;
  timeframe: string;
  strategy: string;
  ticks: number;
  errors: number;
  startedAt: number;
  lastTickAt: number;
  lastPrice: number;
  positionSide: Exposure;
  positionQty: number;
  avgPrice: number;
  unrealized: number;
  realizedToday: number;
  realizedTotal: number;
  ordersToday: number;
  equity: number;
  lastError: string;
}

export interface TickResult {
  signal?: Signal;
  order?: Order;
  fill?: Fill;
  rejected?: string;
  note?: string;
}

type Listener = () => void;

export class TradingEngine {
  private marketData: MarketDataSource;
  private strategy: Strategy;
  private risk: RiskEngine;
  private venue: PaperVenue;
  private position!: Position;

  private timer: ReturnType<typeof setTimeout> | null = null;
  private running = false;
  private killed = false;
  private paused = false;
  private ticking = false;

  private ticks = 0;
  private errors = 0;
  private consecutiveErrors = 0;
  private startedAt = 0;
  private lastTickAt = 0;
  private lastPrice = 0;
  private lastError = '';
  private equityCache = 0;
  private realizedToday = 0;
  private realizedTotal = 0;
  private ordersToday = 0;

  private listeners = new Set<Listener>();

  private constructor(
    private settings: EngineSettings,
    private readonly journal: Journal,
  ) {
    this.marketData = createMarketData(settings.exchange);
    this.strategy = new EmaCrossStrategy(settings.fastPeriod, settings.slowPeriod);
    this.risk = new RiskEngine(settings.risk);
    this.venue = new PaperVenue(settings.startingCash, settings.feePct, settings.slippagePct);
  }

  static async create(settings: EngineSettings, journal: Journal): Promise<TradingEngine> {
    const engine = new TradingEngine(settings, journal);
    await engine.restore();
    return engine;
  }

  /**
   * Rebuild live state from the journal.
   *
   * Called on every start, because on Android "the app was killed" is a normal
   * event rather than a crash. Coming back up must re-read reality, never
   * assume a flat book.
   */
  private async restore(): Promise<void> {
    this.position = await this.journal.rebuildPosition(this.settings.symbol);
    const cash = await this.journal.rebuildCash(this.settings.startingCash);
    this.venue.seed(this.settings.symbol, this.position.quantity, cash);

    this.killed = await this.journal.getControl(KILL_KEY, false);
    this.paused = await this.journal.getControl(PAUSE_KEY, false);
    this.realizedToday = await this.journal.realizedPnl(today());
    this.realizedTotal = await this.journal.realizedPnl();
    this.ordersToday = await this.journal.orderCount(today());
    this.equityCache = this.venue.equity(this.lastPrice || this.position.avgPrice || 0);
  }

  // -- listeners -------------------------------------------------------

  subscribe(listener: Listener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private emit(): void {
    for (const l of this.listeners) {
      try {
        l();
      } catch {
        // A broken listener must never stop the trading loop.
      }
    }
  }

  // -- control ---------------------------------------------------------

  get isRunning(): boolean {
    return this.running;
  }

  async start(): Promise<void> {
    if (this.running) return;
    this.running = true;
    this.startedAt = Date.now();
    this.consecutiveErrors = 0;
    await this.journal.log('info', `engine started — ${this.settings.symbol} ${this.settings.timeframe}`);
    this.emit();
    void this.loop();
  }

  async stop(): Promise<void> {
    this.running = false;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    await this.journal.log('info', 'engine stopped');
    this.emit();
  }

  async kill(reason = 'manual'): Promise<void> {
    this.killed = true;
    await this.journal.setControl(KILL_KEY, true);
    await this.journal.log('warn', `KILL SWITCH ENGAGED (${reason})`);
    this.emit();
  }

  async revive(): Promise<void> {
    this.killed = false;
    await this.journal.setControl(KILL_KEY, false);
    await this.journal.log('info', 'kill switch released');
    this.emit();
  }

  async pause(): Promise<void> {
    this.paused = true;
    await this.journal.setControl(PAUSE_KEY, true);
    this.emit();
  }

  async resume(): Promise<void> {
    this.paused = false;
    await this.journal.setControl(PAUSE_KEY, false);
    this.emit();
  }

  /** Apply new settings. Stops the loop; the caller restarts it. */
  async applySettings(next: EngineSettings): Promise<void> {
    const wasRunning = this.running;
    await this.stop();
    this.settings = next;
    this.marketData = createMarketData(next.exchange);
    this.strategy = new EmaCrossStrategy(next.fastPeriod, next.slowPeriod);
    this.risk = new RiskEngine(next.risk);
    this.venue = new PaperVenue(next.startingCash, next.feePct, next.slippagePct);
    await this.restore();
    if (wasRunning) await this.start();
    this.emit();
  }

  async resetHistory(): Promise<void> {
    const wasRunning = this.running;
    await this.stop();
    await this.journal.reset();
    this.ticks = 0;
    this.errors = 0;
    this.lastError = '';
    await this.restore();
    if (wasRunning) await this.start();
    this.emit();
  }

  // -- the loop --------------------------------------------------------

  private async loop(): Promise<void> {
    while (this.running) {
      try {
        await this.tick();
      } catch (error) {
        this.errors += 1;
        this.consecutiveErrors += 1;
        this.lastError = error instanceof Error ? error.message : String(error);
        await this.journal.log('error', `tick failed: ${this.lastError}`);
        if (this.consecutiveErrors >= 10) {
          await this.kill('too many consecutive errors');
          await this.stop();
          break;
        }
      }
      this.emit();
      if (!this.running) break;
      await this.sleep(this.settings.pollSeconds * 1000);
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => {
      this.timer = setTimeout(resolve, ms);
    });
  }

  /** Run exactly one pass of the pipeline. Exposed for tests. */
  async tick(): Promise<TickResult> {
    if (this.ticking) return { note: 'already ticking' };
    this.ticking = true;
    try {
      this.ticks += 1;
      this.lastTickAt = Date.now();

      const candles = await this.marketData.fetchCandles(
        this.settings.symbol,
        this.settings.timeframe,
        Math.max(this.strategy.warmup + 50, 100),
      );
      if (candles.length === 0) return { note: 'no candles' };

      this.lastPrice = candles[candles.length - 1].close;
      this.equityCache = this.venue.equity(this.lastPrice);
      this.consecutiveErrors = 0;
      this.lastError = '';

      const signal = this.strategy.evaluate(candles);
      if (!signal) return { note: 'strategy abstained' };

      const planned = this.plan(signal);
      if (!planned) return { signal, note: 'target already held' };
      const { order, isExit } = planned;

      const state = await this.riskState(candles[candles.length - 1]);
      const decision = this.risk.assess(signal, this.position, state, isExit);
      if (!decision.approved) {
        await this.journal.recordOrder(order, 'rejected', decision.reason);
        return { signal, order, rejected: decision.reason };
      }

      const sized: Order = { ...order, quantity: decision.quantity };

      let fill: Fill;
      try {
        fill = await this.venue.submit(sized);
      } catch (error) {
        const message = error instanceof ExecutionError ? error.message : String(error);
        this.errors += 1;
        await this.journal.recordOrder(sized, 'failed', message);
        await this.journal.log('error', `execution failed: ${message}`);
        return { signal, order: sized, rejected: message };
      }

      await this.commit(fill, sized);
      return { signal, order: sized, fill };
    } finally {
      this.ticking = false;
    }
  }

  /** Derive the order that moves the live position to the signal's target. */
  private plan(signal: Signal): { order: Order; isExit: boolean } | null {
    const current = this.position.exposure;
    if (current === signal.target) return null;

    const clientId = newClientId();

    // Leaving an open position always takes priority over entering a new one.
    // A reversal is executed as an exit now and an entry next tick, so a single
    // order never crosses through zero.
    if (!this.position.isFlat) {
      return {
        order: {
          symbol: this.settings.symbol,
          side: current === 'long' ? 'sell' : 'buy',
          quantity: Math.abs(this.position.quantity),
          price: signal.price,
          reason: `exit ${current}: ${signal.reason}`,
          clientId,
        },
        isExit: true,
      };
    }

    if (signal.target === 'flat') return null;

    return {
      order: {
        symbol: this.settings.symbol,
        side: signal.target === 'long' ? 'buy' : 'sell',
        quantity: 0, // the risk engine sets the real size
        price: signal.price,
        reason: `enter ${signal.target}: ${signal.reason}`,
        clientId,
      },
      isExit: false,
    };
  }

  private async riskState(candle: Candle): Promise<RiskState> {
    return {
      equity: this.venue.equity(candle.close),
      realizedPnlToday: await this.journal.realizedPnl(today()),
      ordersToday: await this.journal.orderCount(today()),
      killed: this.killed,
      paused: this.paused,
    };
  }

  private async commit(fill: Fill, order: Order): Promise<void> {
    const realized = this.position.apply(fill);
    await this.journal.recordFill(fill, realized);
    await this.journal.recordOrder(order, 'filled', `realized ${realized.toFixed(4)}`);
    await this.journal.log(
      'info',
      `${fill.side.toUpperCase()} ${fill.quantity.toFixed(8)} @ ${fill.price.toFixed(4)} → ${realized >= 0 ? '+' : ''}${realized.toFixed(4)}`,
    );

    this.realizedToday = await this.journal.realizedPnl(today());
    this.realizedTotal = await this.journal.realizedPnl();
    this.ordersToday = await this.journal.orderCount(today());
    this.equityCache = this.venue.equity(this.lastPrice);
  }

  // -- reporting -------------------------------------------------------

  status(): EngineStatus {
    return {
      running: this.running,
      killed: this.killed,
      paused: this.paused,
      mode: 'paper',
      venue: this.venue.name,
      exchange: this.settings.exchange,
      symbol: this.settings.symbol,
      timeframe: this.settings.timeframe,
      strategy: this.strategy.name,
      ticks: this.ticks,
      errors: this.errors,
      startedAt: this.startedAt,
      lastTickAt: this.lastTickAt,
      lastPrice: this.lastPrice,
      positionSide: this.position.exposure,
      positionQty: this.position.quantity,
      avgPrice: this.position.avgPrice,
      unrealized: this.position.unrealizedPnl(this.lastPrice),
      realizedToday: this.realizedToday,
      realizedTotal: this.realizedTotal,
      ordersToday: this.ordersToday,
      equity: this.equityCache,
      lastError: this.lastError,
    };
  }

  getSettings(): EngineSettings {
    return this.settings;
  }

  getJournal(): Journal {
    return this.journal;
  }

  async candles(limit = 90): Promise<Candle[]> {
    return this.marketData.fetchCandles(this.settings.symbol, this.settings.timeframe, limit);
  }
}

File: myp1/runner.py
"""The trading loop.

This is the only place components meet. It owns no strategy logic, no risk
rules and no venue detail — it moves data across the seams in a fixed order:

    market data -> strategy -> position diff -> risk -> execution -> journal -> notify

Every ordering decision in that pipeline exists for a reason. Risk sits
between the strategy and the venue so no signal can reach an exchange
unchecked, and the journal is written before notification so a crash cannot
lose a fill that already happened.
"""

from __future__ import annotations

import asyncio
import logging
import time
import uuid
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Protocol

from .config import Config
from .core.models import Candle, Exposure, Fill, Order, Position, Side, Signal
from .execution.base import ExecutionError, ExecutionVenue
from .marketdata.base import MarketDataSource
from .risk.engine import KILL_KEY, PAUSE_KEY, RiskEngine, RiskState
from .state.journal import Journal, today
from .strategy.base import Strategy

log = logging.getLogger(__name__)

Notifier = Callable[[str], Awaitable[None]]


class SupportsSeed(Protocol):
    def seed_inventory(self, symbol: str, quantity: float, cash: float) -> None: ...


@dataclass
class TickResult:
    """What one pass of the loop did, returned for tests and diagnostics."""

    signal: Signal | None = None
    order: Order | None = None
    fill: Fill | None = None
    rejected: str = ""
    note: str = ""


class TradingRunner:
    def __init__(
        self,
        config: Config,
        *,
        market_data: MarketDataSource,
        strategy: Strategy,
        risk: RiskEngine,
        venue: ExecutionVenue,
        journal: Journal,
        notifier: Notifier | None = None,
    ) -> None:
        self.config = config
        self.market_data = market_data
        self.strategy = strategy
        self.risk = risk
        self.venue = venue
        self.journal = journal
        self.notifier = notifier

        self.symbol = config.symbol
        self.position: Position = journal.rebuild_position(self.symbol)
        self.last_price: float = 0.0
        self.last_tick_at: int = 0
        self.started_at: int = int(time.time() * 1000)
        self.ticks: int = 0
        self.errors: int = 0
        self._running = False

        self._restore_venue_state()

    # -- lifecycle ------------------------------------------------------

    def _restore_venue_state(self) -> None:
        """Rebuild simulator inventory from journalled fills after a restart."""
        seed = getattr(self.venue, "seed_inventory", None)
        if not callable(seed) or self.venue.is_live:
            return
        fills = self.journal.fill_count()
        if fills == 0:
            return
        cash = self.config.starting_cash
        for row in self.journal.recent_fills(limit=10_000):
            notional = row["quantity"] * row["price"]
            cash += (-notional if row["side"] == Side.BUY.value else notional) - row["fee"]
        seed(self.symbol, self.position.quantity, cash)
        log.info(
            "restored paper state from %d fills: qty=%.8f cash=%.2f",
            fills, self.position.quantity, cash,
        )

    @property
    def killed(self) -> bool:
        return bool(self.journal.get_control(KILL_KEY, False))

    @property
    def paused(self) -> bool:
        return bool(self.journal.get_control(PAUSE_KEY, False))

    def kill(self, reason: str = "manual") -> None:
        self.journal.set_control(KILL_KEY, True)
        log.warning("KILL SWITCH ENGAGED (%s)", reason)

    def revive(self) -> None:
        self.journal.set_control(KILL_KEY, False)
        log.warning("kill switch released")

    def pause(self) -> None:
        self.journal.set_control(PAUSE_KEY, True)

    def resume(self) -> None:
        self.journal.set_control(PAUSE_KEY, False)

    # -- one pass -------------------------------------------------------

    async def tick(self) -> TickResult:
        """Run exactly one pass of the pipeline."""
        self.ticks += 1
        self.last_tick_at = int(time.time() * 1000)

        candles = await self.market_data.fetch_candles(
            self.symbol, self.config.timeframe, limit=max(self.strategy.warmup + 50, 100)
        )
        if not candles:
            return TickResult(note="no candles")

        self.last_price = candles[-1].close

        signal = self.strategy.evaluate(candles)
        if signal is None:
            return TickResult(note="strategy abstained")
        self.journal.record_signal(signal)

        order, is_exit = self._plan(signal)
        if order is None:
            return TickResult(signal=signal, note="target already held")

        state = await self._risk_state(candles[-1])
        decision = self.risk.assess(signal, self.position, state, is_exit=is_exit)
        if not decision.approved:
            self.journal.record_order(order, "rejected", decision.reason)
            log.info("order rejected: %s", decision.reason)
            return TickResult(signal=signal, order=order, rejected=decision.reason)

        order = Order(
            symbol=order.symbol,
            side=order.side,
            quantity=decision.quantity,
            price=order.price,
            reason=order.reason,
            client_id=order.client_id,
        )

        try:
            fill = await self.venue.submit(order)
        except ExecutionError as exc:
            self.errors += 1
            self.journal.record_order(order, "failed", str(exc))
            log.error("execution failed: %s", exc)
            await self._notify(f"⚠️ Order failed: {exc}")
            return TickResult(signal=signal, order=order, rejected=str(exc))

        slippage = self.risk.check_slippage(order.price, fill.price)
        if not slippage.approved:
            # The fill already happened; this cannot undo it. Record it,
            # then stop trading so a human looks at why prices are moving
            # this far between signal and execution.
            log.error("slippage breach on filled order: %s", slippage.reason)
            await self._commit(fill, order)
            self.kill(f"slippage breach: {slippage.reason}")
            await self._notify(f"🛑 Kill switch engaged — {slippage.reason}")
            return TickResult(signal=signal, order=order, fill=fill, rejected=slippage.reason)

        await self._commit(fill, order)
        return TickResult(signal=signal, order=order, fill=fill)

    def _plan(self, signal: Signal) -> tuple[Order | None, bool]:
        """Derive the order that moves the live position to the signal target."""
        current = self.position.exposure
        target = signal.target
        if current is target:
            return None, False

        client_id = uuid.uuid4().hex[:12]

        # Leaving an open position always takes priority over entering a new
        # one. A reversal is executed as an exit now and an entry next tick,
        # so a single order never crosses through zero.
        if not self.position.is_flat:
            side = Side.SELL if current is Exposure.LONG else Side.BUY
            return Order(
                symbol=self.symbol,
                side=side,
                quantity=abs(self.position.quantity),
                price=signal.price,
                reason=f"exit {current.value}: {signal.reason}",
                client_id=client_id,
            ), True

        if target is Exposure.FLAT:
            return None, False

        side = Side.BUY if target is Exposure.LONG else Side.SELL
        return Order(
            symbol=self.symbol,
            side=side,
            quantity=0.0,  # risk engine sets the real size
            price=signal.price,
            reason=f"enter {target.value}: {signal.reason}",
            client_id=client_id,
        ), False

    async def _risk_state(self, candle: Candle) -> RiskState:
        try:
            equity = await self.venue.equity(candle.close)
        except ExecutionError as exc:
            log.error("equity lookup failed: %s", exc)
            equity = 0.0  # denies new entries; exits stay permitted
        return RiskState(
            equity=equity,
            realized_pnl_today=self.journal.realized_pnl(today()),
            orders_today=self.journal.order_count(today()),
            killed=self.killed,
            paused=self.paused,
        )

    async def _commit(self, fill: Fill, order: Order) -> None:
        realized = self.position.apply(fill)
        self.journal.record_fill(fill, realized)
        self.journal.record_order(order, "filled", f"realized {realized:.4f}")
        arrow = "🟢" if fill.side is Side.BUY else "🔴"
        await self._notify(
            f"{arrow} {fill.side.value.upper()} {fill.quantity:.8f} {fill.symbol}\n"
            f"price {fill.price:.4f} · fee {fill.fee:.4f}\n"
            f"realized {realized:+.4f} · position {self.position.quantity:.8f}"
        )

    async def _notify(self, text: str) -> None:
        if self.notifier is None:
            return
        try:
            await self.notifier(text)
        except Exception:  # notification must never break the trading loop
            log.exception("notifier failed")

    # -- long-running loop ----------------------------------------------

    async def run(self, *, max_ticks: int | None = None) -> None:
        self._running = True
        mode = "LIVE" if self.config.is_live else "paper"
        log.info("runner started in %s mode on %s %s via %s",
                 mode, self.symbol, self.config.timeframe, self.venue.name)
        await self._notify(
            f"▶️ MY-P1 started — {mode} · {self.symbol} {self.config.timeframe} "
            f"· {self.strategy.name}"
        )

        completed = 0
        try:
            while self._running:
                if max_ticks is not None and completed >= max_ticks:
                    break
                try:
                    await self.tick()
                except asyncio.CancelledError:
                    raise
                except Exception:
                    self.errors += 1
                    log.exception("tick failed")
                    if self.errors >= 10:
                        self.kill("too many consecutive errors")
                        await self._notify("🛑 Kill switch engaged — repeated tick failures")
                        break
                completed += 1
                if max_ticks is None or completed < max_ticks:
                    await asyncio.sleep(self.config.poll_seconds)
        finally:
            self._running = False
            log.info("runner stopped after %d ticks", completed)

    def stop(self) -> None:
        self._running = False

    # -- reporting ------------------------------------------------------

    def status(self) -> dict:
        uptime = (int(time.time() * 1000) - self.started_at) / 1000
        return {
            "mode": "live" if self.config.is_live else "paper",
            "venue": self.venue.name,
            "symbol": self.symbol,
            "timeframe": self.config.timeframe,
            "strategy": self.strategy.name,
            "running": self._running,
            "killed": self.killed,
            "paused": self.paused,
            "ticks": self.ticks,
            "errors": self.errors,
            "uptime_seconds": round(uptime, 1),
            "last_price": self.last_price,
            "position_qty": self.position.quantity,
            "position_side": self.position.exposure.value,
            "avg_price": self.position.avg_price,
            "unrealized": self.position.unrealized_pnl(self.last_price),
            "realized_today": self.journal.realized_pnl(today()),
            "realized_total": self.journal.realized_pnl(),
            "orders_today": self.journal.order_count(today()),
        }
